codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java

```diff
@@ -143,27 +143,30 @@ public void readFrame(ChannelHandlerContext ctx, ByteBuf input, Http2FrameListen
         }
         try {
             do {
-                if (readingHeaders) {
-                    processHeaderState(input);
-                    if (readingHeaders) {
-                        // Wait until the entire header has arrived.
-                        return;
-                    }
+                if (readingHeaders && !preProcessFrame(input)) {
+                    return;
                 }
-
+                verifyHeaderState();
                 // The header is complete, fall into the next case to process the payload.
                 // This is to ensure the proper handling of zero-length payloads. In this
                 // case, we don't want to loop around because there may be no more data
                 // available, causing us to exit the loop. Instead, we just want to perform
                 // the first pass at payload processing now.
-                processPayloadState(ctx, input, listener);
-                if (!readingHeaders) {
-                    // Wait until the entire payload has arrived.
+
+                // Wait until the entire payload has been read.
+                if (input.readableBytes() < payloadLength) {
                     return;
                 }
+                // Slice to work only on the frame being read
+                ByteBuf framePayload = input.readSlice(payloadLength);
+                processPayloadState(ctx, framePayload, listener);
```

Comment:
Some of the verifications result in a stream error, not a connection error. If that happens I think we end up with a corrupt stream. To fix it we can simply move verification to after we've called input.readSlice(..) so that we've consumed the frame from the framing layer. We've already committed to reading frames this large so it doesn't seem like a potential attack vector. The only downside seems to be that for connection errors it can potentially delay closing the session.

Reply:
Agree, I forgot that we aren't skipping the bytes anymore like the original solution if this happens. I'll move this verification after slicing